packages/@uppy/aws-s3/src/MiniXHRUpload.js

```diff
@@ -0,0 +1,360 @@
+          const body = opts.getResponseData(xhr.responseText, xhr)
+          const uploadURL = body[opts.responseUrlFieldName]
+
+          const uploadResp = {
+            status: ev.target.status,
+            body,
+            uploadURL
+          }
+
+          this.uppy.emit('upload-success', file, uploadResp)
+
+          if (uploadURL) {
+            this.uppy.log(`Download ${file.name} from ${uploadURL}`)
+          }
+
+          return resolve(file)
+        } else {
+          const body = opts.getResponseData(xhr.responseText, xhr)
+          const error = buildResponseError(xhr, opts.getResponseError(xhr.responseText, xhr))
+
+          const response = {
+            status: ev.target.status,
+            body
+          }
+
+          this.uppy.emit('upload-error', file, error, response)
+          return reject(error)
+        }
+      })
+
+      xhr.addEventListener('error', (ev) => {
+        this.uppy.log(`[AwsS3/XHRUpload] ${id} errored`)
+        timer.done()
+        queuedRequest.done()
+        if (this.uploaderEvents[file.id]) {
+          this.uploaderEvents[file.id].remove()
+          this.uploaderEvents[file.id] = null
+        }
+
+        const error = buildResponseError(xhr, opts.getResponseError(xhr.responseText, xhr))
+        this.uppy.emit('upload-error', file, error)
+        return reject(error)
+      })
+
+      xhr.open(opts.method.toUpperCase(), opts.endpoint, true)
+      // IE10 does not allow setting `withCredentials` and `responseType`
+      // before `open()` is called.
+      xhr.withCredentials = opts.withCredentials
+      if (opts.responseType !== '') {
+        xhr.responseType = opts.responseType
+      }
+
+      Object.keys(opts.headers).forEach((header) => {
+        xhr.setRequestHeader(header, opts.headers[header])
+      })
+
+      const queuedRequest = this.requests.run(() => {
+        xhr.send(data)
+        return () => {
+          timer.done()
+          xhr.abort()
+        }
+      }, { priority: 1 })
+
+      this._onFileRemoved(file.id, () => {
+        queuedRequest.abort()
+        reject(new Error('File removed'))
+      })
+
+      this._onCancelAll(file.id, () => {
+        queuedRequest.abort()
+        reject(new Error('Upload cancelled'))
+      })
+    })
+  }
+
+  _uploadRemoteFile (file, current, total) {
+    const opts = this._getOptions(file)
+    return new Promise((resolve, reject) => {
+      this.uppy.emit('upload-started', file)
+
+      const fields = {}
+      const metaFields = Array.isArray(opts.metaFields)
+        ? opts.metaFields
+        // Send along all fields by default.
+        : Object.keys(file.meta)
+
+      metaFields.forEach((name) => {
+        fields[name] = file.meta[name]
+      })
+
+      const Client = file.remote.providerOptions.provider ? Provider : RequestClient
+      const client = new Client(this.uppy, file.remote.providerOptions)
+      client.post(file.remote.url, {
```

Comment:
should we also be sending `httpMethod` here?

Reply:
yes i'll apply the recent xhrupload changes here!